src/OpenTelemetry/Trace/Configuration/TracerFactory.cs

```diff
@@ -0,0 +1,229 @@
+﻿// <copyright file="TracerFactory.cs" company="OpenTelemetry Authors">
+// Copyright 2018, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Trace.Configuration
+{
+    using System;
+    using System.Collections.Generic;
+    using OpenTelemetry.Context.Propagation;
+    using OpenTelemetry.Resources;
+    using OpenTelemetry.Trace.Export;
+    using OpenTelemetry.Trace.Sampler;
+
+    public class TracerFactory : TracerFactoryBase, IDisposable
+    {
+        private readonly object lck = new object();
+        private readonly Dictionary<TracerRegistryKey, ITracer> tracerRegistry = new Dictionary<TracerRegistryKey, ITracer>();
+        private readonly List<IDisposable> disposables = new List<IDisposable>();
+
+        private TracerConfigurationOptions tracerConfigurationOptions;
+        private ISampler sampler;
+        private Func<SpanExporter, SpanProcessor> processorFactory;
+        private SpanExporter spanExporter;
+        private SpanProcessor spanProcessor;
+        private IBinaryFormat binaryFormat = new BinaryFormat();
+        private ITextFormat textFormat = new TraceContextFormat();
+
+        private List<Collector> collectorFactories;
+        private ITracer defaultTracer;
+        private bool isBuilt;
+        
+        public TracerFactory SetSampler(ISampler sampler)
+        {
+            this.sampler = sampler;
+            return this;
+        }
+
+        public TracerFactory SetExporter(SpanExporter spanExporter)
+        {
+            this.spanExporter = spanExporter;
+            return this;
+        }
+
+        public TracerFactory SetProcessor(Func<SpanExporter, SpanProcessor> processorFactory)
+        {
+            this.processorFactory = processorFactory;
+            return this;
+        }
+
+        public TracerFactory AddCollector<TCollector>(
+            Func<ITracer, TCollector> collectorFactory)
+            where TCollector : class
```

Comment:
Should there be a base type for collectors? Right now they can be anything, and then optionally implement IDisposable.

Reply:
may be. I don't think they have anything in common today except they all must accept tracer in constructor and they all likely need to be Disposable. I'd prefer to not introduce base collector for now.